src/hotspot/share/opto/rangeinference.cpp

```diff
@@ -0,0 +1,889 @@
+//   being [lo, uhi] and [ulo, hi], lo and uhi are < 0 while ulo and hi are >= 0.
+// This class deals with each interval with both bounds being >= 0 or < 0 in
+// the signed domain. We call it Simple because a canonicalized TypeIntPrototype
+// may contain 1 or 2 SimpleCanonicalResult.
+template <class U>
+class SimpleCanonicalResult {
+  static_assert(U(-1) > U(0), "bit info should be unsigned");
+public:
+  bool _present; // whether this is an empty set
+  RangeInt<U> _bounds;
+  KnownBits<U> _bits;
+
+  bool empty() const {
+    return !_present;
+  }
+
+  static SimpleCanonicalResult<U> make_empty() {
+    return {false, {}, {}};
+  }
+};
+
+// Find the minimum value that is not less than lo and satisfies bits. If there
+// does not exist one such number, the calculation will overflow and return a
+// value < lo.
+//
+// Here, we view a number in binary as a bit string. As a result,  the first
+// bit refers to the highest bit (the MSB), the last bit refers to the lowest
+// bit (the LSB), a bit comes before (being higher than) another if it is more
+// significant, and a bit comes after (being lower than) another if it is less
+// significant.
+template <class U>
+static U adjust_lo(U lo, const KnownBits<U>& bits) {
+  // Violation of lo with respects to bits
+  // E.g: lo    = 1100
+  //      zeros = 0100
+  //      ones  = 1001
+  // zero_violation = 0100, i.e the second bit should be zero, but it is 1 in
+  // lo. Similarly, one_violation = 0001, i.e the last bit should be one, but
+  // it is 0 in lo. These make lo not satisfy the bit constraints, which
+  // results in us having to find the smallest value that satisfies bits
+  U zero_violation = lo & bits._zeros;
+  U one_violation = ~lo & bits._ones;
+  if (zero_violation == one_violation) {
+    // This means lo does not violate bits, it is the result
+    assert(zero_violation == U(0), "");
+    return lo;
+  }
+
+  /*
+  1. Intuition:
+  Call res the lowest value not smaller than lo that satisfies bits, consider
+  the first bit in res that is different from the corresponding bit in lo,
+  since res is larger than lo the bit must be 0 in lo and 1 in res. Since res
+  must satisify bits the bit must be 0 in zeros. Finally, as res should be the
+  smallest value, this bit should be the last one possible.
+
+  E.g:      1 2 3 4 5 6
+       lo = 1 0 0 1 1 0
+        x = 1 0 1 0 1 0
+        y = 0 1 1 1 1 1
+  x would be larger than lo since the first different bit is the 3rd one,
+  while y is smaller than lo because the first different bit is the 1st bit.
+  Next, consider:
+       x1 = 1 0 1 0 1 0
+       x2 = 1 0 0 1 1 1
+  Both x1 and x2 are larger than lo, but x1 > x2 since its first different
+  bit from lo is the 3rd one, while with x2 it is the 7th one. As a result,
+  if both x1 and x2 satisfy bits, x2 would be closer to our true result.
+
+  2. Formality:
+  Call i the largest value such that (with v[0] being the first bit of v, v[1]
+  being the second bit of v and so on):
```

Comment:
Why not move the `v[0]` etc comment up to the `MSB` / `LSB` explanation about first and last bits?